### PR TITLE
Velocyto

### PR DESCRIPTION
**Pull request by tucano**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`

### CODE REVIEW

File: tests/config/test_data.config [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 params {
     // Base directory for test data
-    test_data_base = "https://raw.githubusercontent.com/nf-core/test-datasets/modules"
+    test_data_base = "https://raw.githubusercontent.com/tucano/test-datasets/modules"
```
Comment (sateeshperi) [inline tests/config/test_data.config:6]: this can't be changed....has to point to official nf-core test datasets repo

Comment (tucano) [reply tests/config/test_data.config:6]: fixed ... sorry my bad.

File: modules/nf-core/velocyto/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            run("SAMTOOLS_SORT") {
+                script "../../samtools/sort/main.nf"
+                config "modules/nf-core/velocyto/tests/test.config"
+                
+                process {                                        
+
+                    """
+                    input[0] = [
+                        [ id:'Sample_X' ], // meta map
+                        file(params.test_data['mus_musculus']['velocity']['bam'], checkIfExists: true)
```
Comment (famosab) [inline modules/nf-core/velocyto/tests/main.nf.test:25]: Can you update the filepaths according to #5281 ? :)

File: modules/nf-core/velocyto/tests/main.nf.test [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+
+    // mask location file(params.test_data['mus_musculus']['velocity']['mask'], checkIfExists: true)
+    // test('mus_musculus - barcodes and bam and mask') {
+
+    // }
+
+    
+
+    // test('mus_musculus - barcodes and bam - stub') {
```
Comment (famosab) [inline modules/nf-core/velocyto/tests/main.nf.test:126]: Why is your stub test commented out?

File: modules/nf-core/velocyto/tests/main.nf.test [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                """
+            }
+        }
+        then {
+            assertAll(
+            { assert process.success },
+            {
+                assert snapshot(
+                    process.out.versions
+                ).match()
```
Comment (famosab) [inline modules/nf-core/velocyto/tests/main.nf.test:61]:
```suggestion
                ).match("version")
```

File: modules/nf-core/velocyto/tests/main.nf.test [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            }
+        }
+
+        then {
+            assertAll(
+            { assert process.success },
+            {
+                assert snapshot(
+                    process.out.versions
+                ).match()
```
Comment (famosab) [inline modules/nf-core/velocyto/tests/main.nf.test:100]:
```suggestion
                ).match("presorted_version")
```

**Review by famosab — COMMENTED** [review]
I just left a few minor comments :)

File: modules/nf-core/velocyto/tests/main.nf.test (hunk 2 above)
Comment (tucano) [reply modules/nf-core/velocyto/tests/main.nf.test:25]: Thanks for review this!

Paths updated to new base modules_testdata_base_path. Using

```
params {
    modules_testdata_base_path = "https://raw.githubusercontent.com/tucano/test-datasets/modules/data/"
}
```

FOr testing and looks good (is my fork)

Comment (famosab) [reply modules/nf-core/velocyto/tests/main.nf.test:25]: Ah on the final PR it should look a bit like
```
file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true),
```

File: modules/nf-core/velocyto/tests/main.nf.test (hunk 3 above)
Comment (tucano) [reply modules/nf-core/velocyto/tests/main.nf.test:126]: Sorry for that stub restored.

File: modules/nf-core/velocyto/tests/main.nf.test (hunk 4 above)
Comment (tucano) [reply modules/nf-core/velocyto/tests/main.nf.test:61]: snapshot renamed

File: modules/nf-core/velocyto/tests/main.nf.test (hunk 5 above)
Comment (tucano) [reply modules/nf-core/velocyto/tests/main.nf.test:100]: snapshot renamed

File: modules/nf-core/velocyto/main.nf [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    velocyto run $args -e ${meta.id} -b ${barcodes} -o . ${bam} ${gtf}
+
+    cp .command.out ${prefix}.velocyto.log
```
Comment (asp8200) [inline modules/nf-core/velocyto/main.nf:29]: That seems somewhat unconventional to me. I think 
```
velocyto run $args -e ${meta.id} -b ${barcodes} -o . ${bam} ${gtf} > ${prefix}.velocyto.log
```
or something like that, would resemble more what I see in other modules.

**Comment by tucano** [comment]
Actually I have 3 tests unrelated to my modules in the merged branch and this test are failing.

- nf-test (subworkflows/fastq_fastqc_umitools_fastp, conda)
- nf-test (subworkflows/fastq_fastqc_umitools_fastp, docker)
- nf-test (subworkflows/fastq_fastqc_umitools_fastp, singularity)

What I can do to fix this? I have tested locally with:

`nf-test test subworkflows/nf-core/preprocess_rnaseq --profile docker`

and test looks ok.

Error on github actions are related to nf-test snapshot:

`java.lang.RuntimeException: A snapshot with id 'skip_fastqc' already exists. Snapshot ids have to be unique, and a test can only have one unnamed snapshot.`

File: modules/nf-core/velocyto/meta.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/meta-schema.json
+name: "velocyto"
+description: |
+  Velocyto is a library for the analysis of RNA velocity. velocyto.py CLI use
+  Path(resolve_path=True) and breaks the nextflow logic of symbolic links.
+  If in the work dir velocyto find a file named EXACTLY cellsorted_[ORIGINALBAMNAME]
+  it will skip the samtools sort step. Cellsorted bam file should be cell sorted with:
+    samtools sort -t CB -O BAM -o cellsorted_input.bam input.bam
```
Comment (mashehu) [inline modules/nf-core/velocyto/meta.yml:9]:
````suggestion
```bash
    samtools sort -t CB -O BAM -o cellsorted_input.bam input.bam
```
````

File: modules/nf-core/velocyto/README.md
Comment (mashehu) [inline modules/nf-core/velocyto/README.md:1]: we don't render these readme files anywhere on the nf-core website, which is one of the main ways for people to interact with documentation of modules. Better to have all this info in the meta.yml (as you partially did)

Comment (tucano) [reply modules/nf-core/velocyto/README.md:1]: Added info in the meta description 

5e44e1f64284a60cfdf69bee2d2e3c172f90e179

**Comment by tucano** [comment]
Given the issues with nf-test snapshot related to fastq_fastqc_umitools_fastp subworkflows test I will close this PR and open a new one with a new branch from master.

**Pull request closed without merging**